### ISSUE TITLE
Network request logging

### ISSUE THREAD

**Issue by JaviSoto** [Flipboard/FLEX#5]
I'm thinking about implementing this myself. I believe having a UI where you can check which requests have recently happened and some info about them (status code, headers, etc) would be incredible.

The obvious approach would be to use an `NSURLProtocol` to capture all HTTP(s) requests, store the info about each request in disk / memory, but then return that request back to `NSURLConnection` so that it happens normally.

The tricky thing is that only one `NSURLProtocol` will get to capture any given HTTP request. As soon as one returns `YES` from `-canInitWithRequest:`, the other registered ones won't be consulted (*Or so I believe. I'm starting to doubt myself right now*)

I have an implementation that I could open source of a generic `NSURLProtocol` that lets multiple peers hook into requests (or simply get notified of them) with a middleware type of API (the header is already available on gist: https://gist.github.com/JaviSoto/66763cfacac897d0d27f)

The problem is that if we added this to `FLEX`, and people add `FLEX` to their app with an already existent `NSURLProtocol`, either FLEX's request logging wouldn't work, or we would make the 3rd party's app `NSURLProtocol` stop working.

### Posible Solution
The request logging could be opt-in. `FLEXManager` could have a `-enableRequestLogging` API, that when called, installs the `NSURLProtocol`. Then we could document that the feature wouldn't work in the cases mentioned above.

What do you think about this? Did you have any ideas about how to implement this feature?
Thanks!

**Comment by ryanolsonk** [comment]
I've started working on this. I expect to have something to show within the next few weeks.

**Comment by ryanolsonk** [comment]
I've pushed the progress on this to a branch: https://github.com/Flipboard/FLEX/tree/network-debugging
Feedback is welcome! Here's a GIF of how things currently look:

![requests](https://cloud.githubusercontent.com/assets/1422245/6224592/fef37e88-b632-11e4-9fd2-f788c1a76c07.gif)


I've been using the PonyDebugger approach but I'm not totally satisfied with it. My main problem is that it misses requests made using `+[NSURLConnection sendAsynchronousRequest:...]` or the synchronous variant. I'm considering using an `NSURLProtocol` subclass which would get all requests (as @JaviSoto  suggested), but that comes with some drawbacks too. If we intercept requests with our own `NSURLProtocol`, the way we end up handling each request may be subtly different than it would've been otherwise.

I think I'll take a crack at the `NSURLProtocol` approach since the implementation should be pretty straightforward. I could see offering both options, though it would probably be better if one emerged as the superior approach. If anyone has other ideas on how to hook into the foundation networking stack, I'd love to hear them. I've been digging around in CFNetwork's private headers, but I haven't found a good place to quietly observe request activity. The system `NSURLProtocol` subclasses (e.g. `_NSURLHTTPProtocol`) do nearly all their work outside of Objective-C land where we can't easily swizzle to observe. Hooking the C/C++ functions is probably out of the question for this library.

Aside from that, here's my remaining task list for the feature:
- [ ] Settings menu for the network history list
 - [ ] Clear/trash button to dump recorded requests
 - [ ] Enable switch for request recording.
 - [ ] Slider for recorder cache size (value persisted in `NSUserDefaults`)
- [ ] Sticky header in network history list to indicate when network capture is turned off and how many requests are in the list/accumulated download size.	
- [ ] Test interaction with PonyDebugger
- [ ] Form upload data in request detail view
- [ ] Test in other apps

And some nice to have tasks:
- [ ] Sensitive field redacting
- [ ] Find a way to determine loads from NSURLCache and indicate that in the interface
- [ ] Copy/export option
- [ ] Don’t block the main thread while prettifying responses to push into the response detail view (JSON, plist, etc.).

**Comment by Legoless** [comment]
Does this also support NSURLSession task requests?

**Comment by ryanolsonk** [comment]
The current approach (PonyDebugger) only works with a subset of `NSURLSessionTask` requests. The session must be configured to use a delegate, since PonyDebugger swizzles the delegate methods to do its observation. Similar to the issue with `NSURLConnection`, the async convenience methods do not get detected.

A `NSURLProtocol` based approach could capture all `NSURLSessionTask` requests except for background session tasks.

**Comment by JaviSoto** [comment]
The biggest problem with `NSURLProtocol` is interfering with app's own protocols: when a protocol *captures* a request, it won't be passed on to the next protocol. I've learned that all the SDKs that do any type of network request monitoring, all go down the swizzling path. I'm curious why your approach misses some requests. Did you have troubles swizzling some of the methods? If that's the case, I can help with that, I used to mess a lot with the Obj-C runtime :)

As a side note, I did this in a previous app by monitoring the requests *outside* of it, by using apitools.com and https://github.com/JaviSoto/JSTAPIToolsURLProtocol

**Comment by ryanolsonk** [comment]
TL;DR I think there are good reasons to continue investigating both approaches.

Certainly we could swizzle `+[NSURLConnection sendAsynchronousRequest:queue:completionHandler]`, `+[NSURLConnection sendSynchronousRequest:returningResponse:error]`, and their `NSURLSession` friends, but you do give up some information with that approach. In particular, you loose the detailed timing info (i.e. latency vs data transfer) and the details of any redirects that occurred. That might be an acceptable loss, but it would be awesome if we could get the same (complete) information for all requests.

`NSURLProtocol` is a way we could get consistent logging no mater how a request is made. We can actually do this while still making use of any protocols the app has registered itself. When our protocol receives a request to handle, we will use NSURLConnection or NSURLSession internally to actually send out the request. The registered protocols will get consulted again (this time our protocol will decline to handle it), and the request will be handled by the protocol that would've been used in our absence. Something like this is happening in your library. The requests made from inside the protocol through NSURLConnection will typically make it down to the system's `_NSURLHTTPProtocol`, unless there's a different protocol that can handle the request registered in front of that.

The main trick is that we need to be the first protocol consulted in order to hear about all requests. This applies both in the global list used by `NSURLConnection` and `[NSURLSession sharedSession]`, and in the protocol lists used by any custom `NSURLSession`s. For that, I'm planning to do some swizzling to ensure we're in the first spot in the list.

My main concern with `NSURLProtocol` is that we'll loose something about how the request was intended to be made. For example, we'll have lost the `NSURLSession` context by the time an `NSURLSessionTask` makes it to our protocol. If we use a new `NSURLSession` to handle the task, the `NSURLSessionConfiguration` may be different. Luckily, `NSURLSessionTask` has a private back-pointer to its owning `NSURLSession` that we can use to copy the configuration. However, there could be similar pieces of context that would get dropped with this approach.

One idea that came to mind while writing this is that we could do a more invasive swizzle of methods like `+[NSURLConnection sendAsynchronousRequest:queue:completionHandler]` and actually redirect them to the delegate variants (with ourselves as the delegate) rather than calling through to the original methods. This may prove to be less disruptive than the `NSURLProtocol` approach while providing complete information for all requests.

I'm going to keep exploring both options. Sorry to drone on so long here...

**Comment by ryanolsonk** [comment]
This is now available on master. Check it out and let me know what you think!

**Comment by JaviSoto** [comment]
Just checked it out, it looks amazing :)

**Comment by tijoinc** [comment]
Yay!

**Comment by DaidoujiChen** [comment]
try it :+1:
